overlord/state/dot/dot.go

```diff
@@ -0,0 +1,269 @@
+		if !ok {
+			clusters = append(clusters, lanes)
+		}
+		clusterTasks[clulabel] = append(incluster, t)
+		taskToCluster[t] = clulabel
+	}
+	sort.Slice(clusters, func(i, j int) bool {
+		return lanesLess(clusters[i], clusters[j])
+	})
+	var def []gfrag
+	addToDef := func(f gfrag) {
+		def = append(def, f)
+	}
+	for _, clu := range clusters {
+		clulabel := clusterLabel(clu)
+		addToDef(gcluster{clulabel})
+		for _, t := range clusterTasks[clulabel] {
+			addToDef(gnode{labels[t]})
+		}
+		addToDef(gclusterEnd{})
+	}
+	for _, t := range tasks {
+		clu := taskToCluster[t]
+		haltTasks := t.HaltTasks()
+		sortTasks(haltTasks, labels)
+		for _, t2 := range haltTasks {
+			attrs := ""
+			if taskToCluster[t2] != clu {
+				// cross cluster
+				attrs = "style=bold"
+			}
+			addToDef(&gedge{
+				from:  labels[t],
+				to:    labels[t2],
+				attrs: attrs,
+			})
+		}
+	}
+	tags := []string{chg.Kind()}
+	if tag != "" {
+		tags = append(tags, tag)
+	}
+	return &ChangeGraph{
+		tags: tags,
+		def:  def,
+	}, nil
+}
+
+// String returns a canonical text for the graphviz dot representation of the
+// change dependency graph. This can be used for comparison in tests with some
+// care, OTOH it is not a syntactically complete dot file.
+func (g *ChangeGraph) String() string {
+	buf := new(strings.Builder)
+	const withAttrs = false
+	for _, f := range g.def {
+		f.appendTo(buf, withAttrs)
+	}
+	return strings.TrimSuffix(buf.String(), "\n")
+}
+
+// WriteDotTo writes a syntactically complete dot file for representation of
+// the change dependency graph to the given io.Writer.
+func (g *ChangeGraph) WriteDotTo(w io.Writer) error {
+	if _, err := io.WriteString(w, "digraph {\n"); err != nil {
+		return err
+	}
+	if _, err := fmt.Fprintf(w, "label=\"%s\"\n", strings.Join(g.tags, " - ")); err != nil {
+		return err
+	}
+	const withAttrs = true
+	for _, f := range g.def {
+		if err := f.appendTo(w, withAttrs); err != nil {
+			return err
+		}
+	}
+	_, err := io.WriteString(w, "}\n")
+	return err
+}
+
+// Show invokes the dot command and uses xdg-open to graphically display the
+// graphviz representation of the change dependency graph.
+// If provided logfer is used to report errors and information, if nil stderr
+// and stdout are used instead.
+func (g *ChangeGraph) Show(logfer interface {
+	Logf(format string, args ...interface{})
+}) *ChangeGraph {
+	f, err := os.CreateTemp("", fmt.Sprintf("%s-*.svg", strings.Join(g.tags, "-")))
+	fprintfln := func(w io.Writer, format string, args ...interface{}) {
+		if logfer != nil {
+			logfer.Logf(format, args...)
+		} else {
+			fmt.Fprintf(w, format+"\n", args...)
+		}
+	}
+	if err != nil {
+		fprintfln(os.Stderr, "cannot create .svg file: %v", err)
+		return g
+	}
+	output := f.Name()
+	f.Close()
+	dotCmd := exec.Command("dot", "-Tsvg", "-o"+output)
+	gbuf := new(bytes.Buffer)
+	g.WriteDotTo(gbuf)
+	dotCmd.Stdin = gbuf
+	if o, err := dotCmd.CombinedOutput(); err != nil {
+		if _, ok := err.(*exec.Error); ok {
```

Comment:
Just to confirm, the goal here is to just exit silently if a `dot` executable can't be found right? Since exec.Error is a LookPath error